.github/copilot-instructions.md

```diff
@@ -0,0 +1,837 @@
+    THANDLE_ASSIGN(MY_OBJECT)(&collection->objects[collection->count], object);
+    collection->count++;
+}
+
+void cleanup_collection(OBJECT_COLLECTION* collection)
+{
+    // Clean up all handles
+    for (size_t i = 0; i < collection->count; i++)
+    {
+        THANDLE_ASSIGN(MY_OBJECT)(&collection->objects[i], NULL);  // Decrements ref count
+    }
+    free(collection->objects);
+    collection->objects = NULL;
+    collection->count = 0;
+}
+```
+
+#### Thread Safety and Shared Objects
+```c
+// THANDLE provides thread-safe reference counting
+// Multiple threads can safely share THANDLE instances
+
+typedef struct SHARED_CONTEXT_TAG
+{
+    THANDLE(MY_OBJECT) shared_object;
+    // ... other fields
+} SHARED_CONTEXT;
+
+// Worker thread function
+DWORD WINAPI worker_thread(LPVOID param)
+{
+    SHARED_CONTEXT* context = (SHARED_CONTEXT*)param;
+    
+    // Create local reference to shared object
+    THANDLE(MY_OBJECT) local_ref;
+    THANDLE_ASSIGN(MY_OBJECT)(&local_ref, context->shared_object);
+    
+    // Use object safely - it won't be freed while this thread holds a reference
+    if (local_ref != NULL)
+    {
+        // ... work with local_ref->data ...
+    }
+    
+    // Manual cleanup required before thread exits
+    THANDLE_ASSIGN(MY_OBJECT)(&local_ref, NULL);
+    return 0;
+}
+```
+
+### THANDLE vs Raw Pointers vs REFCOUNT
+
+| Feature | Raw Pointer | REFCOUNT | THANDLE |
+|---------|-------------|----------|---------|
+| **Memory Management** | Manual | Semi-automatic | Automatic |
+| **Type Safety** | None | Some | Strong |
+| **Null Safety** | Prone to errors | Some protection | Built-in checks |
+| **Move Semantics** | N/A | Manual | Built-in |
+| **Scope Safety** | Manual cleanup | Manual cleanup | Manual cleanup |
+| **Thread Safety** | None | Atomic ref counting | Atomic ref counting |
+| **Performance** | Fastest | Fast | Slight overhead |
+
+### Common THANDLE Patterns
+
+#### Factory Pattern with THANDLE
+```c
+// Factory function returning THANDLE
+THANDLE(DATABASE_CONNECTION) database_connection_create(const char* connection_string)
+{
+    THANDLE(DATABASE_CONNECTION) result = THANDLE_MALLOC(DATABASE_CONNECTION)(dispose_function);
+    if (result != NULL)
+    {
+        if (initialize_connection(result, connection_string) != 0)
+        {
+            THANDLE_ASSIGN(DATABASE_CONNECTION)(&result, NULL);  // Clean up on failure
+        }
+    }
+    return result;
+}
+
+// Usage
+THANDLE(DATABASE_CONNECTION) db = database_connection_create("server=localhost");
+if (db != NULL)
+{
+    execute_query(db, "SELECT * FROM users");
+    // Manual cleanup required
+    THANDLE_ASSIGN(DATABASE_CONNECTION)(&db, NULL);
+}
+```
+
+#### Resource Pool with THANDLE
+```c
+typedef struct CONNECTION_POOL_TAG
+{
+    THANDLE(DATABASE_CONNECTION)* connections;
+    size_t pool_size;
+    // ... synchronization primitives
+} CONNECTION_POOL;
+
+THANDLE(DATABASE_CONNECTION) pool_acquire_connection(CONNECTION_POOL* pool)
```

Comment:
>CONNECTION_POOL [](http://example.com/codeflow?start=53&length=15)

any structure (ANY) that has a THANDLE(SOMETHING) in it should be a THANDLE in itself.

Otherwise, ref counting becomes... by convention.

Reply:
Make the CONNECTION_POOL THANDLE and added guidance to have THANDLE be stored in THANDLEs